### PR TITLE
fix: checkpoint handling for exception event

### PR DESCRIPTION
**Pull request by pietro-tota**
#### List of Changes
Add handling for queue consumer processing errors: if an error occurs processing an event it's written into `dead-letter` queue for further analysis. The exception that causes the error is logged in order to discover the routing cause for the event processing error.
The queue checkpoint + dead letter forward logic has been centralize into commons class so that all queue consumer use share can apply the same logic

#### Motivation and Context
Those modifications are needed in order to forward event that cannot be process successfully into dead-letter queue instead of being losed

#### How Has This Been Tested?
Junit tests + local tests with ecommerce local.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt [hunk 1]
```diff
@@ -391,3 +396,40 @@ fun notificationRefundTransactionPipeline(
...
+  // parse the event as a TransactionActivatedEvent just to extract transactionId and event code
+  val binaryData = BinaryData.fromBytes(eventPayload)
+  val event = binaryData.toObject(TransactionActivatedEvent::class.java)
+  val eventLogString = "${event.eventCode}, transactionId: ${event.transactionId}"
+  return checkPointer
+    .success()
+    .doOnSuccess { logger.info("Checkpoint performed successfully for event $eventLogString") }
+    .doOnError { logger.error("Error performing checkpoint for event $eventLogString", it) }
+    .then(pipeline)
+    .then()
```
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt:415]: I add this then() to make the `runPipelineWithDeadLetterQueue` return a Mono<Void> that only replay success/error from the run pipeline + enqueue dead letter event queue. In this way only errors from the dead letter event write are propagated. Errors from pipeline execution are just logged and used to trigger event forwarding into dead letter queue

File: .env.example [hunk 2]
```diff
@@ -5,6 +5,7 @@ TRANSACTIONS_REFUND_QUEUE_NAME=pagopa-ecommerce-transactions-refund-queue
 TRANSACTIONS_CLOSE_PAYMENT_QUEUE_NAME=pagopa-ecommerce-transactions-close-payment-queue
 TRANSACTIONS_NOTIFICATIONS_RETRY_QUEUE_NAME=pagopa-ecommerce-transactions-notifications-retry-queue
 TRANSACTIONS_NOTIFICATIONS_QUEUE_NAME=pagopa-ecommerce-transactions-notifications-queue
+TRANSACTIONS_DEAD_LETTER_QUEUE_NAME=pagopa-ecommerce-transactions-dead-letter-queue
```
[the remark this answers is not in the document]
Comment (pietro-tota) [reply .env.example:8]: Fixed with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/88/commits/1b1dbf005b711546ced22452a80a7a17e419e223

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt [hunk 3]
```diff
@@ -391,3 +396,40 @@ fun notificationRefundTransactionPipeline(
         refundRetryService)
     }
 }
+
+fun eventPipelineCheckpoint(
+  checkPointer: Checkpointer,
+  pipeline: Mono<*>,
```
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt]: Fixed with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/88/commits/1b1dbf005b711546ced22452a80a7a17e419e223

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt [hunk 4]
```diff
@@ -391,3 +396,40 @@ fun notificationRefundTransactionPipeline(
         refundRetryService)
     }
 }
+
+fun eventPipelineCheckpoint(
```
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/common.kt]: Fixed with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/88/commits/1b1dbf005b711546ced22452a80a7a17e419e223

**Review by giovanniberti — APPROVED** [review]

**Pull request merged**